File: BackEnd/src/domain/dtos/plants/create-plant-request.dto.ts
/**
 * @file create-plant-request.dto.ts
 * @description DTO de creación de planta. Validado con class-validator.
 * Compromiso TFG: decoradores de validación en DTO de dominio para simplificar la arquitectura.
 * @module Plants
 * @layer Domain
 */

import {
  IsString, IsNotEmpty, IsOptional, IsInt, IsNumber, Min, Max,
  IsIn, IsBoolean, MaxLength,
} from 'class-validator';
import { Type } from 'class-transformer';

/**
 * DTO de creación de planta recibido del cliente.
 * La foto se gestiona por separado mediante el endpoint POST /upload/image.
 */
export class CreatePlantRequestDTO {
  /** Nombre personalizado de la planta */
  @IsString()
  @IsNotEmpty()
  @MaxLength(100)
  name!: string;

  /** Id de la especie asociada (obligatorio) */
  @IsString()
  @IsNotEmpty()
  speciesId!: string;

  /** URL de imagen en Cloudinary (obtenida previamente de POST /upload/image) */
  @IsOptional()
  @IsString()
  photo?: string;

  /** Ubicación de la planta */
  @IsIn(['Interior', 'Exterior'])
  location!: 'Interior' | 'Exterior';

  /** Ciudad/municipio donde se encuentra la planta (del catálogo de capitales) */
  @IsString()
  @IsNotEmpty()
  @MaxLength(200)
  plantLocation!: string;

  /** Latitud geográfica de la ubicación de la planta */
  @IsOptional()
  @IsNumber()
  @Type(() => Number)
  plantLocationLat?: number;

  /** Longitud geográfica de la ubicación de la planta */
  @IsOptional()
  @IsNumber()
  @Type(() => Number)
  plantLocationLon?: number;

  /** Frecuencia de riego en días */
  @Type(() => Number)
  @IsInt()
  @Min(1)
  @Max(365)
  wateringFrequency!: number;

  /** Necesidad de luz */
  @IsIn(['Low', 'Medium', 'High'])
  lightNeed!: 'Low' | 'Medium' | 'High';

  /** Frecuencia de poda en días (opcional) */
  @IsOptional()
  @Type(() => Number)
  @IsInt()
  @Min(1)
  @Max(365)
  pruningFrequency?: number;

  /** Notas libres */
  @IsOptional()
  @IsString()
  @MaxLength(1000)
  notes?: string;

  /** Si true, el riego considerará datos meteorológicos */
  @IsOptional()
  @IsBoolean()
  considerWeatherForWatering?: boolean;
}

File: FrontEnd/lib/domain/dtos/plants/create_plant_request_dto.dart
/// @file create_plant_request_dto.dart
/// @description DTO de solicitud de creación de planta.
/// Contiene los datos que el usuario envía al crear una nueva planta.
/// @module Plants
/// @layer Domain
library;

/// DTO para crear una nueva planta.
class CreatePlantRequestDto {
  const CreatePlantRequestDto({
    required this.name,
    required this.location,
    required this.plantLocation,
    required this.wateringFrequency,
    required this.lightNeed,
    this.plantLocationLat,
    this.plantLocationLon,
    this.speciesId,
    this.photo,
    this.notes,
  });

  /// Nombre visible de la planta (obligatorio).
  final String  name;

  /// Ubicación: 'Interior' o 'Exterior' (obligatorio).
  final String  location;

  /// Ciudad/municipio donde se encuentra la planta (obligatorio).
  final String  plantLocation;

  /// Frecuencia de riego en días (obligatorio).
  final int     wateringFrequency;

  /// Necesidad de luz: 'Low', 'Medium' o 'High' (obligatorio).
  final String  lightNeed;

  /// Latitud geográfica de la ubicación de la planta (opcional).
  final double? plantLocationLat;

  /// Longitud geográfica de la ubicación de la planta (opcional).
  final double? plantLocationLon;

  /// ID de la especie asociada (opcional).
  final String? speciesId;

  /// URL de la foto (Cloudinary, obtenida previamente via /upload/image).
  final String? photo;

  /// Notas libres del usuario.
  final String? notes;

  /// Serializa el DTO a Map para enviarlo en el cuerpo de la petición HTTP.
  Map<String, dynamic> toJson() => {
    'name':              name,
    'location':          location,
    'plantLocation':     plantLocation,
    'wateringFrequency': wateringFrequency,
    'lightNeed':         lightNeed,
    if (plantLocationLat != null) 'plantLocationLat': plantLocationLat,
    if (plantLocationLon != null) 'plantLocationLon': plantLocationLon,
    if (speciesId != null) 'speciesId': speciesId,
    if (photo     != null) 'photo':     photo,
    if (notes     != null) 'notes':     notes,
  };
}
